### PR TITLE
feat(new-rule): Add WCAG 2.2 target-size rule

### PR DESCRIPTION
**Pull request by WilcoFiers**
- [x] Better handle rounding errors, so that the minimum is not 23.95, but actually 24px
- [x] Tests for target-offset
- [x] Report the offset to the closest neighbour, and indicate the expected offset
- [x] Update the selector to ensure all focusable widgets are included -- may need a matcher too
- [x] Have a matcher filter out inline controls
- [x] Add more integration tests that deal with the selector / matcher
- [x] Document the new check options

Closes issue: #2652

### CODE REVIEW

File: test/commons/dom/visually-sort.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// This method is mostly tested through color-contrast integrations
```
Comment (WilcoFiers) [inline test/commons/dom/visually-sort.js:1]: @straker I could use some help figuring out what needs to be tested here.

File: lib/rules/widget-not-inline-matches.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  (node, vNode) => !svgNamespaceMatches(node, vNode),
+  (node, vNode) => isFocusable(vNode),
+  node => !isInTextBlock(node, { noLengthCompare: true })
+];
+
+function isWidgetType(vNode) {
+  const role = getRole(vNode);
+  if ('option' === role) {
+    return false;
+  } else if (['combobox', 'listbox'].includes(role)) {
```
Comment (WilcoFiers) [inline lib/rules/widget-not-inline-matches.js:21]: For consideration; I could potentially change the role type of these three so that everything we want is a widget, and everything we don't isn't. I didn't explore what the wider implications of that could be.

File: lib/commons/dom/is-in-text-block.js [hunk 3]
```diff
@@ -84,24 +92,27 @@ function isInTextBlock(node) {
...
+  if (options?.noLengthCompare) {
+    return parentText.length !== 0;
+  }
+
+  widgetText = sanitize(widgetText);
+  return parentText.length > widgetText.length;
+}
 
-  return parentText.length > linkText.length;
+function isWidget(domNode) {
```
Comment (straker) [inline lib/commons/dom/is-in-text-block.js:112]: This is the second instance of this function (the other being in `widget-not-inline-matches.js`, but they are slightly different. Thwidget-not-inline-matches.js` only accounts for `listbox` and `combobox` composite roles, but this one includes all composite roles. Should this function be pulled out and made a shared function that both use?

File: test/commons/dom/find-nearby-elms.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      if (vNode.props.id && vNode.props.id !== 'fixture') {
+        ids.push(vNode.props.id);
+      }
+    });
+    return ids;
+  }
+
+  beforeEach(function () {
+    fixture = fixtureSetup(
+      '<div id="n0" style="height:30px; margin-bottom:30px;">0</div>' +
```
Comment (straker) [inline test/commons/dom/find-nearby-elms.js:19]: This test fails when debugging in a browser due to a Mocha element. 

Test result for 'returns node from the same grid cell' showing failure due to 'mocha-stats' node being included in the results

Should probably set the width of these nodes to something small (like 30px) so it doesn't try to include extraneous elements outside of fixture.

**Review by straker — CHANGES_REQUESTED** [review]
I'm not done reviewing this, but figured I get this out there now so you could work on the things I did find.

**Review by straker — CHANGES_REQUESTED** [review]
More things (still working my way through)

File: lib/checks/mobile/target-size.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{
+  "id": "target-size",
+  "evaluate": "target-size-evaluate",
+  "options": {
+    "minSpacing": 24
```
Comment (straker) [inline lib/checks/mobile/target-size.json:5]: `minSpacing` doesn't make much sense for target size. This should be `minSize` or `targetSize`

File: test/checks/mobile/target-size.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    );
+    assert.isTrue(check.evaluate.apply(checkContext, checkArgs));
+    assert.deepEqual(checkContext._data, {
+      minSpacing: 24,
+      width: 40,
+      height: 30
+    });
+  });
+
+  it('returns true for obscured targets with insufficient space', function () {
```
Comment (straker) [inline test/checks/mobile/target-size.js:47]: Should this be `sufficient space`? The test below this one has the same title but a different return.

**Review by straker — DISMISSED** [review]
Finished

File: lib/commons/math/has-visual-overlap.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import { visuallySort } from '../dom';
+
+/**
+ * Check if node A overlaps B
+ * @param {VirtualNode} vNodeA
+ * @param {VirtualNode} vNodeB
+ * @returns 0 | 1 | -1
+ */
+export default function hasVisualOverlap(vNodeA, vNodeB) {
+  const overlapRect = getRectOverlap(
```
Comment (straker) [inline lib/commons/math/has-visual-overlap.js:10]: This could be a much simpler check (and easier to understand) if you used an [axis-aligned bounding box collision check](https://github.com/dequelabs/axe-core/blob/develop/lib/commons/dom/get-rect-stack.js#L129-L137). Then you just check for collision and call `visualSort` if they collide.

```js
return vNodeA.left < vNodeB.left + vNodeB.width &&
    vNodeA.left + vNodeA.width > vNodeB.left &&
    vNodeA.top < vNodeB.top + vNodeB.height &&
    vNodeA.top + vNodeA.height > vNodeB.top &&
    visuallySort(vNodeA, vNodeB) > 0;
```

File: lib/checks/mobile/target-offset-evaluate.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+const roundingMargin = 0.05;
+
+export default function targetOffsetEvaluate(node, options, vNode) {
+  const minOffset = options?.minOffset || 24;
+  const closeNeighbors = [];
+  let closestOffset = minOffset;
+  for (const vNeighbor of findNearbyElms(vNode, minOffset)) {
+    const role = getRole(vNeighbor);
+    if (getRoleType(role) !== 'widget') {
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/checks/mobile/target-offset-evaluate.js:13]: Good point, although not exactly. Disabled buttons should probably be excluded... that's a fair enough point, but I don't think we should exclude inline links for example, or area / SVG widgets. If you put a button too close to your SVG, the button's still the problem, even if axe passes on the SVG.

File: lib/checks/mobile/target-offset.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{
+  "id": "target-offset",
+  "evaluate": "target-offset-evaluate",
+  "options": {
+    "minOffset": 24
+  },
+  "metadata": {
+    "impact": "serious",
+    "messages": {
+      "pass": "Target is sufficiently offset from its closest neighbor (${data.closestOffset}px should be at least ${data.minOffset}px)",
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/checks/mobile/target-offset.json:10]: I put it there because I figured if someone is actually reading these, they're probably trying to figure out why axe-core is passing something they thought should fail. We can do that without putting it in the pass message though. Let me know if you want this removed. It'll still be in `.data`, even if it's not in the message. I can go either way.

File: lib/checks/mobile/target-size.json (hunk 5 above)
Comment (WilcoFiers) [reply lib/checks/mobile/target-size.json:5]: Agreed, should be `minSize`.

File: lib/commons/dom/is-in-text-block.js (hunk 3 above)
Comment (WilcoFiers) [reply lib/commons/dom/is-in-text-block.js:112]: Noticed that did you?... You're right. Need to fix that. I'll go find a better solution. I don't quite like how wonky this is with the listbox / combobox exception. I might just change what type we give to them and avoid needing a function at all.

File: lib/rules/widget-not-inline-matches.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+export default function widgetNotInline(node, vNode) {
+  return matchesFns.every(fn => fn(node, vNode));
+}
+
+const matchesFns = [
+  (node, vNode) => isWidgetType(vNode),
+  (node, vNode) => isNotAreaElement(vNode),
+  (node, vNode) => !svgNamespaceMatches(node, vNode),
+  (node, vNode) => isFocusable(vNode),
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/rules/widget-not-inline-matches.js:13]: Tricky. An element with opacity: 0 is still clickable. All those other on screen visibility methods do prevent focus, but this one doesn't... Also, maybe I need to think about  CSS `pointer-events`, which makes a target non-clickable....

File: lib/commons/dom/create-grid.js [hunk 11]
```diff
@@ -324,11 +324,13 @@ function findScrollRegionParent(vNode, parentVNode) {
...
-  // can find it even if it's in a subgrid
-  vNode._grid = grid;
-
   vNode.clientRects.forEach(rect => {
+    if (rect.right <= 0 || rect.bottom <= 0) {
+      return;
+    }
+    // save a reference to where this element is in the grid so we
+    // can find it even if it's in a subgrid
+    vNode._grid ??= grid;
```
Comment (WilcoFiers) [inline lib/commons/dom/create-grid.js:333]: IMPORTANT: These lines have been modified to fix a bug where elements outside the viewport could still be added to the grid.

File: lib/commons/math/get-offset.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    // Work out which edge of A is farthest away from the center of B
+    const centerB = rectB[start] + rectB[diameter] / 2;
+    const startDistance = Math.abs(centerB - rectA[start]);
+    const endDistance = Math.abs(centerB - rectA[end]);
+    if (startDistance >= endDistance) {
+      farthestPoint[axis] = rectA[start]; // left | top
+    } else if (startDistance < endDistance) {
+      farthestPoint[axis] = rectA[end]; // right | bottom
+    }
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/commons/math/get-offset.js:40]: Why would that be an improvement? Xor'ing a bit on the int is going to be faster than squaring it, no? My version of this (minus the elseif, I agree) better expresses what's happening. Even if yours is faster, which I doubt, it's confusing.

File: lib/commons/math/get-offset.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    adjacentRect = ownRect;
+  }
+
+  const { top, right, bottom, left } = adjacentRect;
+  // Is the adjacent rect horizontally or vertically aligned
+  const xAligned = x >= left && x <= right;
+  const yAligned = y >= top && y <= bottom;
+  // Find the closest edge of the adjacent rect
+  const closestX = Math.abs(left - x) < Math.abs(right - x) ? left : right;
+  const closestY = Math.abs(top - y) < Math.abs(bottom - y) ? top : bottom;
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/commons/math/get-offset.js:72]: As above. I doubt this is faster, but it's definitely not expressing what's going on as well.

File: lib/commons/math/get-offset.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/commons/math/get-offset.js:40]: Going to leave these. Pow is slightly faster, but not enough that it's worth making the code less readable.

File: lib/checks/keyboard/no-focusable-content-evaluate.js [hunk 14]
```diff
@@ -41,8 +41,7 @@ function getFocusableDescendants(vNode) {
 
   const retVal = [];
   vNode.children.forEach(child => {
-    const role = getRole(child);
-    if (getRoleType(role) === 'widget' && isFocusable(child)) {
+    if (getRoleType(child) === 'widget' && isFocusable(child)) {
```
Comment (WilcoFiers) [inline lib/checks/keyboard/no-focusable-content-evaluate.js:44]: If you want me to pull these into their own PR let me know. It's not unrelated to the PR... but it's not very related either.

File: lib/commons/math/has-visual-overlap.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import { visuallySort } from '../dom';
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/commons/math/has-visual-overlap.js:1]: Nope... but this is a good bit confusing. `dom.visuallyOverlaps` looks that a rect intersects with a particular element. It doesn't check whether A is on top of B, or B is on top of A. Probably `dom.visuallyOverlaps` should have been called `rectIntersectsElement` or something like that? Don't think it impacts this PR.

File: lib/commons/math/get-offset.js [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+  dimensionProps.forEach(([axis, start, end, diameter]) => {
+    if (rectB[start] < rectA[start] && rectB[end] > rectA[end]) {
+      farthestPoint[axis] = rectA[start] + rectA[diameter] / 2; // center | middle
+      return;
+    }
+    // Work out which edge of A is farthest away from the center of B
+    const centerB = rectB[start] + rectB[diameter] / 2;
+    const startDistance = Math.abs(centerB - rectA[start]);
+    const endDistance = Math.abs(centerB - rectA[end]);
+    if (startDistance >= endDistance) {
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/commons/math/get-offset.js:38]: If the distance is the same it doesn't matter.

File: test/commons/dom/find-nearby-elms.js (hunk 4 above)
Comment (WilcoFiers) [reply test/commons/dom/find-nearby-elms.js]: How do I get to this? And should we care?

File: lib/rules/widget-not-inline-matches.js [hunk 17]
```diff
@@ -15,14 +15,7 @@ const matchesFns = [
...
 function isWidgetType(vNode) {
-  const role = getRole(vNode);
-  if ('option' === role) {
-    return false;
-  } else if (['combobox', 'listbox'].includes(role)) {
-    return true;
-  }
-  const roleType = getRoleType(role);
-  return roleType === 'widget';
+  return getRoleType(vNode) === 'widget';
```
Comment (straker) [inline lib/rules/widget-not-inline-matches.js:18]: So this will return `true` for the `option` role, is that OK?

File: lib/checks/mobile/target-size-evaluate.js [hunk 18]
```diff
@@ -10,32 +10,44 @@ const roundingMargin = 0.05;
...
+  for (const vNeighbor of findNearbyElms(vNode)) {
+    if (
+      !hasVisualOverlap(vNode, vNeighbor) ||
+      getCssPointerEvents(vNeighbor) === 'none'
+    ) {
+      continue;
+    }
+    if (isEnclosedRect(vNode, vNeighbor)) {
+      this.data({ messageKey: 'obscured' });
+      return true;
```
Comment (straker) [inline lib/checks/mobile/target-size-evaluate.js:30]: Should add the node as a related node

File: test/checks/mobile/target-size.js [hunk 19]
```diff
@@ -39,45 +38,139 @@ describe('target-size tests', function () {
...
+          'display: inline-block; width:40px; height:30px; margin-left:30px;' +
+          '">x</button>' +
+          '<button style="' +
+          'display: inline-block; width:40px; height:30px; margin-left: -10px;' +
+          '">x</button>' +
+          '<span tabindex="0" style="' +
+          'display: inline-block; width:40px; height:30px; margin-left: -100px;' +
+          '">x</span>'
+      );
+      assert.isTrue(check.evaluate.apply(checkContext, checkArgs));
```
Comment (straker) [inline test/checks/mobile/target-size.js:99]: I feel like this should be `false`. the `span` with `tabindex=0` captures the click on the left side of the target, making the clickable width effectively 20px.

image

File: test/checks/mobile/target-size.js [hunk 20]
```diff
@@ -39,45 +38,139 @@ describe('target-size tests', function () {
...
+          'display: inline-block; width:40px; height:30px; margin-left:30px;' +
+          '">x</button>' +
+          '<button style="' +
+          'display: inline-block; width:40px; height:30px; margin-left: -10px;' +
+          '">x</button>' +
+          '<button disabled style="' +
+          'display: inline-block; width:40px; height:30px; margin-left: -100px;' +
+          '">x</button>'
+      );
+      assert.isTrue(check.evaluate.apply(checkContext, checkArgs));
```
Comment (straker) [inline test/checks/mobile/target-size.js:119]: I feel this should be `false` as well since the disabled button captures the click on the left side of the target, making the clickable width effectively 20px. 

This can be debatable though as the disabled button doesn't cause a misclick to happen.

image

**Review by straker — CHANGES_REQUESTED** [review]

File: lib/commons/aria/get-role-type.js [hunk 21]
```diff
@@ -1,21 +1,23 @@
...
 function getRoleType(role) {
-  const roleDef = standards.ariaRoles[role];
-
-  if (!roleDef) {
-    return null;
+  if (
+    role instanceof AbstractVirtualNode ||
+    (window?.Node && role instanceof window.Node)
+  ) {
+    role = axe.commons.aria.getRole(role);
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/commons/aria/get-role-type.js:17]: I didn't noticed that. Have I really? I used instanceof here because that's what we usually do when we want to check nodes. There are lots of `instanceof AbstractVirtualNode` calls.. and a couple `instanceof window.Node` calls too.

Looking at where/how this function has been used, it's going to get `null` passed into it a good bit too, since getRole() can return null. I'll update it to reflect that.

File: lib/commons/dom/is-in-text-block.js [hunk 22]
```diff
@@ -33,10 +34,16 @@ function getBlockParent(node) {
...
+ * widgets. When false, compares the length of non-widget text to widget text
+ *
  * @param  {Element} node [description]
+ * @param  {Object} options Optional
+ * @property {Bool} noLengthCompare
  * @return {Boolean}      [description]
  */
-function isInTextBlock(node) {
+function isInTextBlock(node, options) {
+  const { getRoleType } = axe.commons.aria;
```
[the remark this answers is not in the document]
Comment (WilcoFiers) [reply lib/commons/dom/is-in-text-block.js:46]: Ran into this error again: https://github.com/dequelabs/axe-core/issues/3281

Looks like I can work around it by importing directly from get-role-type, rather than from aria/index. Doing that.

File: lib/rules/widget-not-inline-matches.js (hunk 17 above)
Comment (WilcoFiers) [reply lib/rules/widget-not-inline-matches.js:18]: They do, but they're pulled out for not being focusable.

File: test/checks/mobile/target-size.js (hunk 19 above)
Comment (WilcoFiers) [reply test/checks/mobile/target-size.js:99]: This is no different than if that was just a narrower button. If someone accidentally hits the span instead of the button, nothing happens. Or at least, we can't assume something is going to happen. If that span is a widget that is missing a role, a different accessibility violation exists in that page, and after they've fixed that (by adding a role to the span) this rule will work correctly.

We could maybe report this as incomplete? I'd prefer we do this later though. This PR is much larger than I want it to be already. Please open an issue for this.

File: test/checks/mobile/target-size.js (hunk 20 above)
Comment (WilcoFiers) [reply test/checks/mobile/target-size.js:119]: Same answer. I don't think we can fail this, maybe incomplete, but I'd prefer we leave that for another time.

**Review by straker — APPROVED** [review]

**Review by straker — APPROVED** [review]
